neqo-transport/src/frame.rs

```diff
@@ -263,43 +263,71 @@ impl Frame {
         fin: bool,
         space: usize,
     ) -> Option<(Frame, usize)> {
-        let mut remaining = space.saturating_sub(1 + Encoder::varint_len(stream_id));
+        let mut overhead = 1 + Encoder::varint_len(stream_id);
         if offset > 0 {
-            remaining = remaining.saturating_sub(Encoder::varint_len(offset));
+            overhead += Encoder::varint_len(offset);
+        }
+
+        if overhead > space {
+            qdebug!(
+                "Frame::new_stream -> None; hdrlen {} > space {}",
+                overhead,
+                space
+            );
+            return None;
         }
-        let (fin, fill) = if data.len() > remaining {
-            if remaining == 0 {
-                return None;
-            }
 
+        let (fin, fill) = if data.len() > (space - overhead) {
             // More data than fits, fill the packet and negate |fin|.
             (false, true)
-        } else if data.len() == remaining {
+        } else if data.len() == (space - overhead) {
             // Exact fit, fill the packet, keep |fin|.
             (fin, true)
         } else {
             // Too small, so include a length.
-            let data_len = min(remaining - 1, data.len());
-            remaining -= Encoder::varint_len(u64::try_from(data_len).unwrap());
-            remaining = min(data.len(), remaining);
-            // In case the added length causes this to spill over, check |fin| again.
-            (fin && remaining == data.len(), false)
+            let data_len = min(space - overhead - 1, data.len());
+            overhead += Encoder::varint_len(u64::try_from(data_len).unwrap());
+
+            // If all data isn't going to make it in the frame, don't keep fin.
+            let keep_fin = data.is_empty() || data.len() <= (space - overhead);
```

Comment:
OK, I see now how my suggestion might have been a problem.  Without including the overhead>space check in this block, this check can underflow.  That's not going to show up in tests because you don't have a >16383 length frame that is one shorter than the available space.  That's the only case where the addition of the length causes the addition of a length to overflow the available space.  

You should add a test for a 16384 byte range being added to a space that is one byte longer than needed.  That is a very strange case to deal with, because the addition of the length causes this code to decide not to include all the data, but that causes the length to be encoded on one fewer byte and the frame ends up both not including all the data, but also not using all the remaining space.

You can correct the underflow by changing this test to:
```suggestion
            let keep_fin = data.is_empty() || (data.len() + overhead <= space);
```
But it would be better to make the same change to avoid underflow throughout.  In #374 I learned that clippy::pedantic wants the code to use `match x.cmp(y)` and this could too, as follows:

```rust
let (fin, fill) = match (data.len() + overhead).cmp(&space) {
  Ordering::Greater => (false, true),
  Ordering::Equal => (fin, true),
  Ordering::Lesser => {
    overhead += ...;
    let keep_fin = data.len() + overhead <= space;
    (fin && keep_fin, false)
  }
};
```

In doing so, you can remove the check for `overhead > space` from before this loop, because it will just hit the `Ordering::Lesser` arm.  Then you have two checks after this block:

```rust
if overhead > space {
  qdebug!(...);
  return None;
}
let data_len = min(data.len(), space - overhead);
if data_len == 0 && !fin {
  qdebug!(...);
  return None;
}
```

Note that you can drop the `data.is_empty()` arm on keep_fin.  If there is no data to send and the remaining space is larger than the available length, then we'll only ever add 1 byte more, so the data - at most - will fill the space.